File: google_sheet.rb
require 'google/apis/sheets_v4'
require 'googleauth'
require 'json'
require 'fileutils'

class GoogleSheet
  APPLICATION_NAME = 'LinkedIn Scrabber'
  SERVICE_ACCOUNT_KEY_PATH = './credentials.json'
  attr_reader :config
  
  # Initializer to get the path and save config
  def initialize(file_path)
    @file_path = file_path
    load_config
  end

  private

  # Method to load and parse JSON file
  def load_config
    if File.exist?(@file_path)
      begin
        @config = JSON.parse(File.read(@file_path))
      rescue JSON::ParserError => e
        puts "Failed to parse JSON: #{e.message}"
        @config = {}
      end
    else
      puts "File not found: #{@file_path}"
      @config = {}
    end
  end

  # Authorize using the service account key
  def authorize_service_account
    key_file = File.read(SERVICE_ACCOUNT_KEY_PATH)
    scope = ['https://www.googleapis.com/auth/spreadsheets']

    authorizer = Google::Auth::ServiceAccountCredentials.make_creds(
      json_key_io: StringIO.new(key_file),
      scope: scope
    )
    
    authorizer.fetch_access_token!
    authorizer
  end

  # Initialize the Sheets API
  def initialize_service

    service = Google::Apis::SheetsV4::SheetsService.new
    service.client_options.application_name = APPLICATION_NAME
    service.authorization = authorize_service_account

    service
  end

  # Append an error log to a specific sheet (tab) within the existing spreadsheet
  def add_to_sheet(row, page, service)

    range = "#{page}!A2"
    value_range_object = Google::Apis::SheetsV4::ValueRange.new(
      values: [
        [
        row[:id],
        row[:title],
        row[:company],
        row[:location],
        Time.now.strftime('%Y-%m-%d %H:%M'),
        row[:date],
        "https://www.linkedin.com/jobs/view/#{row[:id]}"
        ]
      ]
    )
    
    service.append_spreadsheet_value(
      config['spreadsheet']['id'],
      range,
      value_range_object,
      value_input_option: 'RAW'
    )
  end

  public def add_to_sheet_if_needed(row, page)
    sleep 1 # to avoid limit
    
    service = initialize_service
    range = "#{page}!A:A" # Define the range to check (first column in #{ENV['SPREADSHEET_PAGE']})

    # Fetch all values from the first column
    response = service.get_spreadsheet_values(config['spreadsheet']['id'], range)

    # Check if the ID exists in the first column
    if response.values.flatten.include?(row[:id].to_s)
      puts "⏭️ Already imported"
    else
      add_to_sheet(row, page, service)
      puts "✅ Added"
    end
  end
end

File: googleSheet.js
const { google } = require('googleapis');
const fs = require('fs').promises;
const path = require('path');

class GoogleSheet {

  constructor(config) {
    this.config = config;
  }

  async init() {
    this.service = await this.initializeService();
    await this.addSheetIfNotExists("LinkedIn");
    await this.addSheetIfNotExists("Indeed");
  }

  async authorizeServiceAccount() {
    // Reading service account credentials
    const keyFilePath = path.resolve(__dirname, 'credentials.json');
    let keyFileData;

    try {
      keyFileData = await fs.readFile(keyFilePath, 'utf8');
    } catch (err) {
      throw new Error(`Service account key file not found: ${keyFilePath}`);
    }

    const parsedKeyFile = JSON.parse(keyFileData);

    const jwtClient = new google.auth.JWT(
      parsedKeyFile.client_email,
      null,
      parsedKeyFile.private_key,
      ['https://www.googleapis.com/auth/spreadsheets']
    );

    await jwtClient.authorize();
    return jwtClient;
  }

  async initializeService() {
    const auth = await this.authorizeServiceAccount();
    const sheets = google.sheets({ version: 'v4', auth });
    return sheets;
  }

  async addToSheetIfNeeded(row, page) {
    // Wait 1 second to avoid rate limits (mimicking original code)
    await new Promise((resolve) => setTimeout(resolve, 1000));

    const spreadsheetId = this.config.spreadsheet?.id;
    if (!spreadsheetId) {
      console.error('Spreadsheet ID not found in config.');
      return;
    }

    const range = `${page}!A:A`;

    const response = await this.service.spreadsheets.values.get({
      spreadsheetId,
      range,
    });

    const existingValues = response.data.values || [];
    const ids = existingValues.flat();
    if (ids.includes(row.id.toString())) {
      console.log('⏭️ Already imported');
    } else {
      await this.addToSheet(row, page);
      console.log('✅ Added');
    }
  }

  async addToSheet(row, page) {
    const spreadsheetId = this.config.spreadsheet?.id;
    if (!spreadsheetId) {
      console.error('Spreadsheet ID not found in config.');
      return;
    }

    const range = `${page}!A2`;
    const values = [
      [
        row.id,
        row.title,
        row.company,
        row.location,
        new Date().toISOString().replace('T', ' ').slice(0, 16),
        row.date || "",
        row.link
      ],
    ];

    await this.service.spreadsheets.values.append({
      spreadsheetId,
      range,
      valueInputOption: 'RAW',
      requestBody: {
        values,
      },
    });
  }

  async addSheetIfNotExists(sheetName) {
    const spreadsheetId = this.config.spreadsheet.id;
    try {
      // Step 1: Get all sheets in the spreadsheet
      const response = await this.service.spreadsheets.get({
        spreadsheetId,
      });

      const existingSheets = response.data.sheets.map(sheet => sheet.properties.title);

      // Step 2: Check if the sheet already exists
      if (existingSheets.includes(sheetName)) {
        console.log(`Sheet "${sheetName}" already exists.`);
        return;
      }

      // Step 3: Add the new sheet
      await this.service.spreadsheets.batchUpdate({
        spreadsheetId,
        requestBody: {
          requests: [
            {
              addSheet: {
                properties: {
                  title: sheetName,
                },
              },
            },
          ],
        },
      });

      console.log(`Sheet "${sheetName}" created successfully.`);
    } catch (error) {
      console.error('Error managing sheets:', error);
    }
  }
}

module.exports = GoogleSheet;